mne/decoding/ssd.py

```diff
@@ -0,0 +1,316 @@
+        if len(ch_types) > 1:
+            raise ValueError('At this point SSD only supports fitting '
+                             'single channel types. Your info has %i types' %
+                             (len(ch_types)))
+        self.info = info
+        self.freqs_signal = (filt_params_signal['l_freq'],
+                             filt_params_signal['h_freq'])
+        self.freqs_noise = (filt_params_noise['l_freq'],
+                            filt_params_noise['h_freq'])
+        self.filt_params_signal = filt_params_signal
+        self.filt_params_noise = filt_params_noise
+        self.sort_by_spectral_ratio = sort_by_spectral_ratio
+        if n_fft is None:
+            self.n_fft = int(self.info['sfreq'])
+        else:
+            self.n_fft = int(n_fft)
+        self.picks_ = (Ellipsis if picks is None else picks)
+        self.return_filtered = return_filtered
+        self.reg = reg
+        self.n_components = n_components
+        self.rank = rank
+        self.cov_method_params = cov_method_params
+
+    def _check_X(self, X):
+        """Check input data."""
+        if not isinstance(X, np.ndarray):
+            raise ValueError('X should be of type ndarray (got %s).'
+                             % type(X))
+        if X.ndim < 2:
+            raise ValueError('X must have at least 2 dimensions.')
+        elif X.ndim > 3:
+            raise ValueError('X must have at most 3 dimensions.')
+
+        n_chan = 0
+        if X.ndim == 2:
+            n_chan = X.shape[0]
+        elif X.ndim == 3:
+            n_chan = X.shape[1]
+        if n_chan != self.info['nchan']:
+            raise ValueError('Info must match the input data.'
+                             'Found %i channels but expected %i.' %
+                             (n_chan, self.info['nchan']))
+
+    def fit(self, X, y=None):
+        """Estimate the SSD decomposition on raw or epoched data.
+
+        Parameters
+        ----------
+        X : array, shape ([n_epochs, ]n_channels, n_times)
+            The input data from which to estimate the SSD. Either 2D array
+            obtained from continuous data or 3D array obtained from epoched
+            data.
+        y : None | array, shape (n_samples,)
+            Used for scikit-learn compatibility.
+
+        Returns
+        -------
+        self : instance of SSD
+            Returns the modified instance.
+        """
+        self._check_X(X)
+        if X.ndim == 2:
+            X_aux = X[self.picks_, :].copy()
+
+        elif X.ndim == 3:
+            X_aux = X[:, self.picks_, :].copy()
+        else:
+            raise NotImplementedError()
+
+        X_signal = filter_data(
+            X_aux, self.info['sfreq'], **self.filt_params_signal)
+        X_noise = filter_data(
+            X_aux, self.info['sfreq'], **self.filt_params_noise)
+
+        X_noise -= X_signal
+        if X.ndim == 3:
+            X_signal = np.hstack(X_signal)
+            X_noise = np.hstack(X_noise)
```

Comment:
Avoid a copy and a conditional by using reshape tricks instead
```suggestion
        X_signal.shape = (-1, X_signal.shape[-1])
        X_noise.shape = (-1, X_noise.shape[-1])
```

Reply:
it comes with an issue though, since when working with 3D arrays `(n_epochs, n_channels, n_times)` this produces a new shape of `(n_epochs*n_channels, n_times)` and no what we want: `(n_channels, n_epochs*n_times)`. when I do `X_signal.shape = (X_signal.shape[1], -1)` it works perfect for 3D arrays but not for the 2D (since it is like transpose it)

Reply:
mmhmm not sure about this. When working with 3D data of shape `(n_epochs, n_channels, n_times)` this reshape will make data of shape `(n_epochs, n_channels*n_times)` instead of what we want: `(n_channels, n_epochs*n_times)`. I don't find a way to use the shape trick when dealing with both 2D and 3D data. or there is something am I missing?

Reply:
Ahh yep my suggestion is wrong. There are transpose + reshape tricks that could be used to avoid the conditional but it will actually make the code less clear in this case, so it's fine to keep this as is!